segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,741 @@
+        logStorage.addToList((int) logStorage.getLength(), ledgerData);
+        return logStorage;
+    }
+
+    private LogStorage fenceAllTheLedgers(String streamSegmentName, LogStorage ledger) throws StreamSegmentException {
+        List<String> children = null;
+        try {
+            children = zkClient.getChildren().forPath(getZkPath(streamSegmentName));
+        } catch (Exception e) {
+            throw translateZKException(streamSegmentName, e);
+        }
+        if (children.size() == 0) {
+            LedgerData ledgerData = createLedgerAt(streamSegmentName, 0);
+            ledger.addToList(0, ledgerData);
+        } else {
+            CompletableFuture<LedgerData>[] futures = null;
+            children.stream().forEach(child -> {
+                try {
+                int offset = Integer.valueOf(child);
+                LedgerData ledgerData = null;
+                    ledgerData = fenceLedgerAt(streamSegmentName, offset);
+                if (ledgerData.getLedgerHandle().getLength() == 0) {
+                    tryDeleteLedger(ledgerData.getLedgerHandle().getId());
+                    zkClient.delete().deletingChildrenIfNeeded().forPath(getZkPath(streamSegmentName) + "/" + child);
+                }
+                ledger.addToList(offset, ledgerData);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+            return ledger;
+    }
+
+    private void tryDeleteLedger(long ledgerId) {
+        bookkeeper.asyncDeleteLedger(ledgerId, (rc, ctx) -> {
+            if (rc != 0) {
+                log.warn("Deletion of ledger {} failed with errorCode {}", ledgerId, rc);
+            }
+        }, null);
+    }
+
+    private LedgerData fenceLedgerAt(String streamSegmentName, int firstOffset) throws Exception {
+        Stat stat = new Stat();
+        byte[] data = zkClient.getData()
+                              .storingStatIn(stat)
+                              .forPath(getZkPath(streamSegmentName) + "/" + firstOffset);
+
+        LedgerData ledgerData = deserializeLedgerData(firstOffset, data, stat);
+        ledgerData.setLength((int) ledgerData.getLedgerHandle().getLength());
+        ledgerData.setLastAddConfirmed(ledgerData.getLedgerHandle().getLastAddConfirmed());
+        log.info("Fencing ledger {}", streamSegmentName);
+        ledgerData.getLedgerHandle().close();
+        return ledgerData;
+    }
+
+    private void writeDataAt(LedgerHandle lh, long offset, InputStream data, int length, String segmentName) {
+        log.info("Writing {} at offset {} for ledger {}", length, offset, lh.getId());
+        byte[] bytes = new byte[length];
+        try {
+            StreamHelpers.readAll(data, bytes, 0, length);
+        } catch (IOException e) {
+            throw new CompletionException(e);
+        }
+
+        Exceptions.handleInterrupted(() -> {
+            try {
+                lh.addEntry(bytes, 0, length);
```

Comment:
This will always fail if *length* is greater than the configured limit on BookKeeper (by default 1MB).
I can't find any upstream mechanism to split the data in smaller chunks.
Do you have any pointer ?